web/src/engine/osk/src/keyboard-layout/oskLayerGroup.ts

```diff
@@ -165,86 +171,97 @@ export default class OSKLayerGroup {
-          bestMatchDistance = distance;
-          row = r;
-        }
+      } else if(distance < bestMatchDistance) {
+        bestMatchDistance = distance;
+        row = r;
       }
     }
-
     // Assertion:  row no longer `null`.
+    // (We already prevented the no-rows available scenario, anyway.)
 
-    // Warning: am not 100% sure that what follows is actually fully correct.
-
-    // Find minimum distance from any key
-    let closestKeyIndex = 0;
-    let dx: number;
-    let dxMax = 24;
-    let dxMin = 100000;
+    // Step 2: Find minimum distance from any key
+    // - If the coord is within a key's square, go ahead and return it.
+    let closestKey: OSKBaseKey = null;
+    let minDistance = Number.MAX_VALUE;
 
-    const x = coord.clientX;
+    for (let key of row.keys) {
+      const keySpec = key.spec;
+      if(keySpec.sp == ButtonClasses.blank || keySpec.sp == ButtonClasses.spacer) {
+        continue;
+      }
 
-    for (let k = 0; k < row.keys.length; k++) {
-      // Second-biggest, though documentation suggests this is probably right.
-      const keySquare = row.keys[k].square as HTMLElement; // gets the .kmw-key-square containing a key
-      const squareRect = keySquare.getBoundingClientRect();
+      // Max distance from the key's center to consider, horizontally.
+      const keyRadius = keySpec.proportionalWidth / 2;
+      const distanceFromCenter = Math.abs(proportionalCoords.x - keySpec.proportionalX);
 
       // Find the actual key element.
-      let childNode = keySquare.firstChild ? keySquare.firstChild as HTMLElement : keySquare;
-
-      if (childNode.className !== undefined
-        && (childNode.className.indexOf('key-hidden') >= 0
-          || childNode.className.indexOf('key-blank') >= 0)) {
-        continue;
-      }
-      const x1 = squareRect.left;
-      const x2 = squareRect.right;
-      if (x >= x1 && x <= x2) {
-        // Within the key square
-        return <KeyElement>childNode;
-      }
-      dx = x1 - x;
-      if (dx >= 0 && dx < dxMin) {
-        // To right of key
-        closestKeyIndex = k; dxMin = dx;
-      }
-      dx = x - x2;
-      if (dx >= 0 && dx < dxMin) {
-        // To left of key
-        closestKeyIndex = k; dxMin = dx;
+      if(distanceFromCenter - keyRadius <= 0) {
+        // As noted above:  if we land within a key's square, match instantly!
+        return key.btn;
+      } else {
+        const distance = distanceFromCenter - keyRadius;
+        if(distance < minDistance) {
+          minDistance = distance;
+          closestKey = key;
+        }
       }
     }
 
-    if (dxMin < 100000) {
-      const t = <HTMLElement>row.keys[closestKeyIndex].square;
-      const squareRect = t.getBoundingClientRect();
+    // Step 3:  If the input coordinate wasn't within any valid key's "square",
+    // determine if the nearest valid key is acceptable.
+    const minHorizFudge = NEAREST_KEY_MIN_HORIZONTAL_FUDGE_PIXELS / this.element.offsetWidth;
```

Comment:
does this force a layout reflow? Shouldn't we be using `this.computedWidth`?

This seems like a funny calculation to me. this.element.offsetWidth is surely going to be a large number like 600px or something right? So then we have 24/600 = 0.04. I don't understand how this will work.

Reply:
> does this force a layout reflow?

We're not changing key layout information - at worst, we're adjusting the styling (via key-highlighting).

> Shouldn't we be using `this.computedWidth`?

Perhaps so. 

> So then we have 24/600 = 0.04.

Exactly!  That lands us with a minimum fudge factor matching 4% of total keyboard width.  If 60% of a key's width is less than that, we use that "4%" as the fudge factor.  The original logic existed to prevent the fudge factor from becoming too small for practical use.

So, taking that _other_ comment into account, what we probably want is...

- A default of ~6% total keyboard width as a "nearest key" fudge factor...
- ... unless that 6% allocates less than 24 pixels - in which case, use 24 pixels as the fudge factor.

Also, in case it's not otherwise clear already, the `minDistance` value seen here is using "%age of keyboard width" coordinates.

Reply:
> does this force a layout reflow?

At the time, I was operating under the (possibly false) assumption that a higher-level element would, at least, induce much less reflow than digging down to the specific keys.  The LayoutGroup element is static as long as the keyboard lives, while the layers and keys hop in and out of the rendering pipeline.

It would be able to trigger a minor reflow if the layer of the key isn't currently the active layer, yes.

Reply:
Righto, that all makes sense thanks.

I wonder if we could use a term other than fudge factor, because I was struggling to correlate that. Perhaps `keyTouchMarginPx` (where it is in pixels) or something like that?

> So, taking that _other_ comment into account, what we probably want is...
> 
> * A default of ~6% total keyboard width as a "nearest key" fudge factor...
> * ... unless that 6% allocates less than 24 pixels - in which case, use 24 pixels as the fudge factor.

Hmm, 6% of total keyboard width is probably good enough? That would be 60% of width of a 'typical' key on a 'typical' 10-key-wide layout.

Pixels is a little fraught due to device resolution differences, it would be better to be a minimum physical distance if we wanted to use that, but not sure that's readily available to us?

Thus why I am wondering if just using the 6% of total width is sufficient.